chainer/datasets/pickle_dataset.py

```diff
@@ -0,0 +1,164 @@
+import threading
+
+import six.moves.cPickle as pickle
+
+from chainer.dataset import dataset_mixin
+
+
+class PickleDatasetWriter(object):
+
+    """Writer class which makes PickleDataset.
+
+    To make :class:`PickleDataset`, a user need to prepare data using
+    :class:`PickleDatasetWriter`.
+
+    Args:
+        writer: File like object that supports ``write`` and ``tell`` methods.
+        protocol (int): Valid protocol for :mod:`pickle`.
+
+    .. seealso: chainer.datasets.PickleDataset
+
+    """
+
+    def __init__(self, writer, protocol=pickle.HIGHEST_PROTOCOL):
+        self._positions = []
+        self._writer = writer
+        self._protocol = protocol
+
+    def close(self):
+        self._writer.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def write(self, x):
+        position = self._writer.tell()
+        pickle.dump(x, self._writer, protocol=self._protocol)
+        self._positions.append(position)
+
+    def flush(self):
+        self._writer.flush()
+
+
+class PickleDataset(dataset_mixin.DatasetMixin):
+
+    """Dataset stored in a storage using pickle.
+
+    :mod:`pickle` is the default serialization library of Python.
+    This dataset stores any objects in a storage using :mod:`pickle`.
+    Even when a user want to use a large dataset, this dataset can stores all
+    data in a large storage like HDD and each data can be randomly accessible.
+
+    >>> with chainer.datasets.open_pickle_dataset_writer('/path/to/data') as w:
+    ...     w.write((1, 2.0, 'hello'))
+    ...     w.write((2, 3.0, 'good-bye'))
+    ...
+    >>> with chainer.datasets.open_pickle_dataset('/path/to/data') as dataset:
+    ...     print(dataset[1])
+    ...
+    (2, 3.0, 'good-bye')
+
+    Args:
+        reader: File like object. `reader` must support random access.
+
+    """
+
+    def __init__(self, reader):
+        # TODO(unnonouno): It is better to check seekable(), but the method is
+        # not supported in py2
```

Comment:
How about checking `seekable` when the Python in use is 3.x, e.g. using `six.PY3`?